python_modules/libraries/dagster-components/dagster_components/core/component_rendering.py

```diff
@@ -0,0 +1,68 @@
+from typing import Any, Mapping, Optional, Sequence, Type, TypeVar, Union
+
+from dagster._record import record
+from jinja2 import Template
+from pydantic import BaseModel
+
+T = TypeVar("T")
+
+REF_BASE = "#/$defs/"
+REF_TEMPLATE = f"{REF_BASE}{{model}}"
+
+
+@record
+class TemplatedValueResolver:
+    context: Mapping[str, Any]
+
+    def resolve(self, val: str) -> str:
+        return Template(val).render(**self.context)
+
+
+def _should_render(
+    valpath: Sequence[Union[str, int]], json_schema: Optional[Mapping[str, Any]]
+) -> bool:
+    """For a given path to an attribute, read the assoicated metadata in the json_schema to determine if the value should be rendered."""
+    if json_schema is None:
+        return True
+
+    subschema = json_schema
+    for el in valpath:
+        if isinstance(el, str):
+            subschema = subschema["properties"].get(el)
+        else:
+            subschema = subschema["items"]
+
+        if not subschema:
+            return False
+        if "$ref" in subschema:
+            subschema = json_schema["$defs"].get(subschema["$ref"][len(REF_BASE) :])
+        if subschema.get("defer_render") is True:
+            return False
+
+    return True
```

Comment:
some comments here would be helpful. This just looks like magical incantations to me as someone that doesn't know a lot about json schema